website/blog/2025-04-08-run-ferretdb-postgres-documentdb-extension-cnpg-kubernetes.md

```diff
@@ -0,0 +1,243 @@
+        B[PostgreSQL Cluster]
+        C[PostgreSQL Pod 1 with DocumentDB extension]
+        D[PostgreSQL Pod 2 with DocumentDB extension]
+        E[PostgreSQL Pod 3 with DocumentDB extension]
+        F[FerretDB]
+        G[FerretDB Pod]
+        H[postgres-cluster-rw Service]
+    end
+
+    A -->|Manages| B
+    B -->|Creates| C
+    B -->|Creates| D
+    B -->|Creates| E
+    F -->|Runs| G
+    G -->|Connects to| H
+    B -->|Exposes| H
+```
+
+## Install CNPG with Helm
+
+Install the CNPG operator using Helm.
+This operator will manage the PostgreSQL cluster.
+
+```sh
+helm repo add cnpg https://cloudnative-pg.github.io/charts
+helm upgrade --install cnpg \
+  --namespace cnpg \
+  --create-namespace \
+  cnpg/cloudnative-pg
+```
+
+## Create the PostgreSQL cluster
+
+We'll now define a PostgreSQL cluster using CNPG.
+This cluster will run 3 Postgres instances using the PostgreSQL with DocumentDB extension image.
+
+Save this as `pg-cluster.yaml`:
+
+```yaml
+apiVersion: postgresql.cnpg.io/v1
+kind: Cluster
+metadata:
+  name: postgres-cluster
+  namespace: cnpg
+spec:
+  instances: 3
+  imageName: 'ghcr.io/ferretdb/postgres-documentdb:17-0.102.0-ferretdb-2.1.0'
+  postgresUID: 999
+  postgresGID: 999
+  enableSuperuserAccess: true
+
+  storage:
+    size: 1Gi
+
+  postgresql:
+    shared_preload_libraries:
+      - pg_cron
+      - pg_documentdb_core
+      - pg_documentdb
+    parameters:
+      cron.database_name: 'postgres'
+
+  bootstrap:
+    initdb:
+      postInitSQL:
+        - 'CREATE EXTENSION IF NOT EXISTS documentdb CASCADE;'
+```
+
+We explicitly enable `enableSuperuserAccess` so that we can connect with the default `postgres` user.
+
+CloudNativePG runs PostgreSQL with UID 26 by default, but the DocumentDB image requires UID and GID 999.
+Be sure to set both explicitly to avoid permission errors.
+
+We also load the required extensions and enable shared preload libraries needed by FerretDB.
+The `cron.database_name` parameter is set to `postgres` to ensure that the cron jobs run in the correct database.
+
+The `bootstrap` section is where we run the SQL command to create the DocumentDB extension.
+Since CloudNativePG doesn't execute entrypoint scripts from the DocumentDB image, we need to manually create the extension during initialization.
+
+Apply it:
+
+```sh
+kubectl apply -f pg-cluster.yaml -n cnpg
+```
+
+CNPG will handle cluster creation, persistent storage, and generate a password for the `postgres` superuser in Secret.
+
+You can check the status of the cluster with:
+
+```sh
+kubectl get cluster -n cnpg
+```
+
+You should see the cluster in `Running` state.
+
+```text
+NAME               AGE    INSTANCES   READY   STATUS                     PRIMARY
+postgres-cluster   176m   3           3       Cluster in healthy state   postgres-cluster-1
+```
+
+To get the generated password for the `postgres` user:
+
+```sh
+kubectl get secret -n cnpg postgres-cluster-superuser -o jsonpath='{.data.password}' | base64 -d && echo
```

Comment:
Why do we need `&& echo`?

Reply:
Not all that essential, actually. But without it, the password is printed directly followed by the shell prompt on the same line (e.g., `password%`). Adding it appends a newline, so the prompt appears without it as just `password`.